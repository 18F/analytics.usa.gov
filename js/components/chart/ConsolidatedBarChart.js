import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";

/**
 * Retrieves the data report from the passed data URL and creates a
 * consolidated bar chart visualization for the data in the provided key.
 *
 * @param {object} props the properties for the component.
 * @param {string} props.dataUrl the URL of the location of the data to be
 * downloaded for the chart.
 * @param {string} props.chartDataKey the key of the data object to use for the chart
 * proportions.
 * @param {number} props.maxItems the max number of items to be displayed in the
 * bar chart after consolidating items beyond the maximum.
 * @returns {import('react').ReactElement} The rendered element
 */
function ConsolidatedBarChart({ dataUrl, chartDataKey, maxItems }) {
  const ref = useRef(null);
  const [chartData, setChartData] = useState(null);

  useEffect(() => {
    const initChart = async () => {
      if (!chartData) {
        const data = await DataLoader.loadJSON(dataUrl);
        await setChartData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder.buildConsolidatedBarchart(
          ref.current,
          chartData,
          chartDataKey,
          maxItems,
        );
      }
    };
    initChart().catch(console.error);
  }, [chartData]);

  return (
    <figure id={`chart_${chartDataKey}`} ref={ref}>
      <div className="data chart__bar-chart"></div>
    </figure>
  );
}

ConsolidatedBarChart.propTypes = {
  dataUrl: PropTypes.string.isRequired,
  chartDataKey: PropTypes.string.isRequired,
  maxItems: PropTypes.number.isRequired,
};

export default ConsolidatedBarChart;
