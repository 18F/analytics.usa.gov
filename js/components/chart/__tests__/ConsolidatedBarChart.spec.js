import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import ConsolidatedBarChart from "../ConsolidatedBarChart";
import DataLoader from "../../../lib/data_loader";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("ConsolidatedBarChart", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <ConsolidatedBarChart
          dataUrl="http://www.example.com/data/"
          chartDataKey="browser"
          maxItems={10}
        />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    describe("and the expected key exists in the loaded data", () => {
      describe("and the data items is greater than the max items param", () => {
        beforeEach(async () => {
          data = {
            name: "browsers",
            query: {},
            meta: {
              name: "Browsers",
              description:
                "30 days of visits broken down by browser for all sites. (>100 sessions)",
            },
            totals: {
              visits: 1492721065,
              by_browser: {
                Chrome: 724669791,
                Safari: 532418559,
                Edge: 131456418,
                "Safari (in-app)": 28370663,
                Firefox: 28673898,
                "(other)": 542273,
                "Samsung Internet": 21535617,
                "Android Webview": 14658083,
                Opera: 3256137,
                "Amazon Silk": 2018466,
                "(not set)": 1299162,
                "Mozilla Compatible Agent": 882647,
                YaBrowser: 499856,
                "Internet Explorer": 1883528,
                "Whale Browser": 123256,
                "Coc Coc": 56684,
                "UC Browser": 119932,
                "Opera Mini": 49514,
                "Aloha Browser": 26440,
                Vivaldi: 14972,
                Mozilla: 27911,
                "Android Runtime": 22723,
                "Android Browser": 9271,
                "DuckDuckGo Browser": 13066,
                PaleMoon: 12977,
                "Meta Quest Browser": 8586,
                "Phoenix Browser": 8795,
                SeaMonkey: 9990,
                Lynx: 38014,
                Waterfox: 2841,
                Iron: 9347,
                Seznam: 1204,
                Puffin: 444,
              },
            },
            taken_at: "2024-03-11T13:59:19.359Z",
          };

          DataLoader.loadJSON.mockImplementation(() => {
            return Promise.resolve(data);
          });

          component = render(
            <ConsolidatedBarChart
              dataUrl="http://www.example.com/data/"
              chartDataKey="browser"
              maxItems={10}
            />,
          );

          await waitFor(() => screen.getByText("Chrome"));
          // Wait for barchart transition animation to complete (200 ms, set in
          // js/lib/chart_helpers/barchart.js)
          await delay(300);
        });

        it("renders a component with data loaded with the max number of items including an 'other' item", () => {
          expect(component.asFragment()).toMatchSnapshot();
        });
      });

      describe("and the data items is less than the max items param", () => {
        beforeEach(async () => {
          data = {
            name: "browsers",
            query: {},
            meta: {
              name: "Browsers",
              description:
                "30 days of visits broken down by browser for all sites. (>100 sessions)",
            },
            totals: {
              visits: 1492721065,
              by_browser: {
                Chrome: 724669791,
                Safari: 532418559,
                Edge: 131456418,
                "Safari (in-app)": 28370663,
                Firefox: 28673898,
                "Samsung Internet": 21535617,
                "Android Webview": 14658083,
                Opera: 3256137,
                "Amazon Silk": 2018466,
              },
            },
            taken_at: "2024-03-11T13:59:19.359Z",
          };

          DataLoader.loadJSON.mockImplementation(() => {
            return Promise.resolve(data);
          });

          component = render(
            <ConsolidatedBarChart
              dataUrl="http://www.example.com/data/"
              chartDataKey="browser"
              maxItems={10}
            />,
          );

          await waitFor(() => screen.getByText("Chrome"));
          // Wait for barchart transition animation to complete (200 ms, set in
          // js/lib/chart_helpers/barchart.js)
          await delay(300);
        });

        it("renders a component with data loaded without an 'other' item", () => {
          expect(component.asFragment()).toMatchSnapshot();
        });
      });
    });

    describe("and the expected key does not exist in the loaded data", () => {
      beforeEach(async () => {
        data = {
          name: "browsers",
          query: {},
          meta: {
            name: "Browsers",
            description:
              "30 days of visits broken down by browser for all sites. (>100 sessions)",
          },
          totals: {
            visits: 1492721065,
            by_os: {
              iOS: 545836253,
              Windows: 446590663,
              Android: 288132572,
              Macintosh: 149049201,
              Linux: 33028818,
              "(other)": 542273,
              "Chrome OS": 28865009,
              "": 24490,
              Tizen: 27195,
              "(not set)": 1272412,
              Fuchsia: 20999,
              "OS/2": 19323,
              "Firefox OS": 15342,
              "Playstation 4": 10575,
              FreeBSD: 3893,
              UNIX: 781,
              BlackBerry: 1397,
            },
          },
          taken_at: "2024-03-11T13:59:19.359Z",
        };

        DataLoader.loadJSON.mockImplementation(() => {
          return Promise.resolve(data);
        });

        component = render(
          <ConsolidatedBarChart
            dataUrl="http://www.example.com/data/"
            chartDataKey="browser"
            maxItems={10}
          />,
        );

        // Wait for barchart transition animation to complete (200 ms, set in
        // js/lib/chart_helpers/barchart.js)
        await delay(300);
      });

      it("renders a component without data", () => {
        expect(component.asFragment()).toMatchSnapshot();
      });
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <ConsolidatedBarChart
          dataUrl="http://www.example.com/data/"
          chartDataKey="browser"
          maxItems={10}
        />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
