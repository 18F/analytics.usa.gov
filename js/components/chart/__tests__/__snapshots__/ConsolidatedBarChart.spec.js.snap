// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsolidatedBarChart when data is loaded and the expected key does not exist in the loaded data renders a component without data 1`] = `
<DocumentFragment>
  <figure
    id="chart_browser"
  >
    <div
      class="data chart__bar-chart margin-top-4"
    />
  </figure>
</DocumentFragment>
`;

exports[`ConsolidatedBarChart when data is loaded and the expected key exists in the loaded data and the data items is greater than the max items param renders a component with data loaded with the max number of items including an 'other' item 1`] = `
<DocumentFragment>
  <figure
    id="chart_browser"
  >
    <div
      class="data chart__bar-chart margin-top-4"
    >
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Chrome
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          48.5%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 49%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Safari
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          35.7%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 36%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Edge
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          8.8%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 9%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Firefox
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          1.9%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 2%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Safari (in-app)
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          1.9%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 2%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Samsung Internet
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          1.4%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 1%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Android Webview
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          1%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 1%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Opera
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          0.2%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 0%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Amazon Silk
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          0.1%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 0%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Other
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          0.4%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 0%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`ConsolidatedBarChart when data is loaded and the expected key exists in the loaded data and the data items is less than the max items param renders a component with data loaded without an 'other' item 1`] = `
<DocumentFragment>
  <figure
    id="chart_browser"
  >
    <div
      class="data chart__bar-chart margin-top-4"
    >
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Chrome
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          48.7%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 49%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Safari
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          35.8%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 36%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Edge
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          8.8%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 9%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Firefox
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          1.9%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 2%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Safari (in-app)
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          1.9%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 2%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Samsung Internet
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          1.4%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 1%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Android Webview
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          1%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 1%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Opera
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          0.2%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 0%;"
        />
      </div>
      <div
        class="chart__bar-chart__item"
      >
        <div
          class="chart__bar-chart__item__label dark-grey text--overflow-ellipsis margin-right-8"
        >
          Amazon Silk
        </div>
        <div
          class="chart__bar-chart__item__value"
        >
          0.1%
        </div>
        <div
          class="chart__bar-chart__item__bar bg-palette-color-5"
          style="width: 0%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`ConsolidatedBarChart when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <figure
    id="chart_browser"
  >
    <div
      class="data chart__bar-chart margin-top-4"
    />
  </figure>
</DocumentFragment>
`;

exports[`ConsolidatedBarChart when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <figure
    id="chart_browser"
  >
    <div
      class="data chart__bar-chart margin-top-4"
    />
  </figure>
</DocumentFragment>
`;
